src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ArrayBuffer.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Net.Http
+{
+    // Warning: Mutable struct!
+    // The purpose of this struct is to simplify buffer management.
+    // It manages a sliding buffer where bytes can be added at the end and removed at the beginning.
+    // [ActiveSpan/Memory] contains the current buffer contents; these bytes will be preserved
+    // (copied, if necessary) on any call to EnsureAvailableBytes.
+    // [AvailableSpan/Memory] contains the available bytes past the end of the current content,
+    // and can be written to in order to add data to the end of the buffer.
+    // Commit(byteCount) will extend the ActiveSpan by [byteCount] bytes into the AvailableSpan.
+    // Discard(byteCount) will discard [byteCount] bytes as the beginning of the ActiveSpan.
+
+    internal struct ArrayBuffer
+    {
+        private byte[] _bytes;
+        private int _activeStart;
+        private int _availableStart;
+
+        // Invariants:
+        // 0 <= _activeStart <= _availableStart <= bytes.Length
+
+        public ArrayBuffer(int initialSize)
+        {
+            _bytes = new byte[initialSize];
+
+            _activeStart = 0;
+            _availableStart = 0;
+        }
+
+        public Span<byte> ActiveSpan => new Span<byte>(_bytes, _activeStart, _availableStart - _activeStart);
+        public Span<byte> AvailableSpan => new Span<byte>(_bytes, _availableStart, _bytes.Length - _availableStart);
+        public Memory<byte> ActiveMemory => new Memory<byte>(_bytes, _activeStart, _availableStart - _activeStart);
+        public Memory<byte> AvailableMemory => new Memory<byte>(_bytes, _availableStart, _bytes.Length - _availableStart);
+
+        public void Discard(int byteCount)
+        {
+            Debug.Assert(byteCount <= ActiveSpan.Length);
+            _activeStart += byteCount;
+
+            if (_activeStart == _availableStart)
+            {
+                _activeStart = 0;
+                _availableStart = 0;
+            }
+        }
+
+        public void Commit(int byteCount)
+        {
+            Debug.Assert(byteCount <= AvailableSpan.Length);
+            _availableStart += byteCount;
+        }
+
+        // Ensure at least [byteCount] bytes to write to.
+        public void EnsureAvailableSpace(int byteCount)
+        {
+            if (byteCount <= AvailableSpan.Length)
+            {
+                return;
+            }
+
+            int totalFree = _activeStart + AvailableSpan.Length;
+            if (byteCount <= totalFree)
+            {
+                // We can free up enough space by just shifting the bytes down, so do so.
```

Comment:
Any value in making it a circular buffer, such that we can wrap around rather than shifting everything down?

Reply:
Yeah, I'm trying to keep it as simple as possible right now.

There are all sorts of potential improvements, including circular buffer, linked buffers, etc. All of them add some complexity to the code that uses this. So I'm deferring any cleverness until we have some data that shows it actually matters.

Reply:
That said, I'm certainly open to making changes now that don't introduce a bunch of new complexity... Including naming etc.